### ISSUE TITLE
ogbn_arxiv.yaml with 32 partitions configuration for training

### ISSUE THREAD

**Issue by jk912** [marius-team/marius#160]
Hi, I am trying to run ogbn_arxiv with MariusGNN with 32 partitions.

I am using this command for preprocessing.
`marius_preprocess --dataset ogbn_arxiv --output_dir arxiv_example/ --num_partitions 32`

I am using this command for training.
`marius_train marius_origin/marius/examples/configuration/ogbn_arxiv.yaml`

First question is: I just realized that if we do not write "num_partitions 32" in the yaml file, I am not using 32 partitions for the training. Am I right?

Second question is: if I need to declare num_partition = 32 in the yaml file. What should I add on the yaml file?

Below is the given ogbn_arxiv.yaml file by Marius Team for your convenience.

> model:
>   learning_task: NODE_CLASSIFICATION
>   encoder:
>     train_neighbor_sampling:
>       - type: ALL
>       - type: ALL
>       - type: ALL
>     layers:
>       - - type: FEATURE
>           output_dim: 128
>           bias: true
>       - - type: GNN
>           options:
>             type: GRAPH_SAGE
>             aggregator: MEAN
>           input_dim: 128
>           output_dim: 128
>           bias: true
>       - - type: GNN
>           options:
>             type: GRAPH_SAGE
>             aggregator: MEAN
>           input_dim: 128
>           output_dim: 128
>           bias: true
>       - - type: GNN
>           options:
>             type: GRAPH_SAGE
>             aggregator: MEAN
>           input_dim: 128
>           output_dim: 40
>           bias: true
>   decoder:
>     type: NODE
>   loss:
>     type: CROSS_ENTROPY
>     options:
>       reduction: SUM
>   dense_optimizer:
>     type: ADAM
>     options:
>       learning_rate: 0.01
> storage:
>   device_type: cuda
>   dataset: 
>     dataset_dir: datasets/ogbn_arxiv_example/
>   edges:
>     type: DEVICE_MEMORY
>     options:
>       dtype: int
>   features:
>     type: DEVICE_MEMORY
>     options:
>       dtype: float
> training:
>   batch_size: 1000
>   num_epochs: 10
>   pipeline:
>     sync: true
> evaluation:
>   batch_size: 1000
>   pipeline:
>     sync: true
>

**Comment by JasonMoho** [comment]
Yes, you need to set the storage of the embeddings to use the partition buffer with 32. See docs: http://marius-project.org/marius/config_interface/configuration.html#partition-buffer

Although running arxiv with 32 partitions doesn't make much practical sense since the dataset is so tiny, partitioning should be used for large datasets where the embeddings don't fit in CPU memory.

**Comment by jk912** [comment]
@JasonMoho Thank you for the response. I have read the papers of Marius and MariusGNN. And I also have read all the issues opened and closed. But I have some questions... I got interests in GNN nowdays and I have no colleagues around me..

Another question is: When CPU memory is 64GB, do you have a recommendation for dataset? Papers100M could be nice, but it is killed. And what factor should I consider when I set the actual number of partitions?

**Comment by JasonMoho** [comment]
Thanks for your interest.

For a smaller dataset that will fit in memory consider [OGBNProducts](https://github.com/marius-team/marius/blob/main/src/python/tools/preprocess/datasets/ogbn_products.py#L13)

Papers100M requires roughly 70GB of CPU memory to use in-memory training. But you can use the partition buffer to scale to disk. See our [config](https://github.com/marius-team/marius/blob/eurosys_2023_artifact_improvements/experiment_manager/system_comparisons/configs/ogbn_papers100m_seq_acc/marius_gs.yaml#L71) for training on papers100m for the MariusGNN paper.

If using disk-based training you can use these quick rules for setting the parameters:
- **num_partitions**: I'd recommend you set this to roughly `num_nodes / 10,000` rounded down to the nearest power of 2. So for papers100m, this is **8192**.
- **buffer_capacity**: This controls the amount of partitions kept in memory. We want to maximize this for best training accuracy. Set this to the largest power of two such that `(buffer_capacity  / num_partitions) * dataset_size < (CPU memory capacity - 16 GB)`. The 16GB is there so we have enough memory left for other data and intermediate operations. So for papers100m and 64GB of CPU memory this will give a buffer capacity of **4096**. Larger dataset sizes or smaller CPU memory capacities will require smaller buffer capacities.
- **fine_to_coarse_ratio**: This is a parameter specific to the COMET ordering, setting it properly helps with model accuracy. Set to `(4 * num_partitions / buffer_capacity) * num_partitions`. So for the above numbers this gives **1024**

These rules should give a decent config for disk-based training as starting point.

There is a more advanced set of tuning rules at the end of section 6 of the MariusGNN paper, which we used to set the config I shared above.

**Comment by jk912** [comment]
@JasonMoho Thank you for the kind explanation even though the questions were not well-explained! It helps a lot..

**Comment by jk912** [comment]
@JasonMoho I just have a last question.  We set partition_buffer for type of storage. What about edges, nodes and features for storage type? When we want to train a dataset with 10 partitions, are we fine to just remain the configuration like below?

storage:
  device_type: cuda
  dataset:
    dataset_dir: /home/data/datasets/ogbn_arxiv/
  edges:
    type: DEVICE_MEMORY
  nodes:
    type: DEVICE_MEMORY
  features:
    type: DEVICE_MEMORY
  embeddings:
    type: PARTITION_BUFFER
    options:
      dtype: float
      num_partitions: 10
      buffer_capacity: 5
      prefetching: true
      type: DEVICE_MEMORY
  prefetch: true
  shuffle_input: true
  full_graph_evaluation: true

**Comment by jk912** [comment]
if we preprocess with 32 partitions and we dont put such sentences like
"features:
    type: PARTITION_BUFFER
    options:
      dtype: float
      num_partitions: 32
      buffer_capacity: 24
"

, we are not training 32 partitions properly?

**Comment by jk912** [comment]
Just figured it out with #147 post
